### PR TITLE
implement hard replay protection

### PR DESCRIPTION
**Pull request by fireice-uk**

### CODE REVIEW

File: src/wallet/wallet2.cpp [hunk 1]
```diff
@@ -3449,7 +3449,14 @@ bool wallet2::sign_tx(const std::string &unsigned_filename, const std::string &s
     tools::wallet2::pending_tx &ptx = signed_txes.ptx.back();
     crypto::secret_key tx_key;
     std::vector<crypto::secret_key> additional_tx_keys;
-    bool r = cryptonote::construct_tx_and_get_tx_key(m_account.get_keys(), m_subaddresses, sd.sources, sd.splitted_dsts, sd.change_dts.addr, sd.extra, ptx.tx, sd.unlock_time, tx_key, additional_tx_keys, sd.use_rct);
+
+    bool use_replay_prot = m_testnet || get_daemon_blockchain_height() >= 137510;
```
Comment (psychocrypt) [inline src/wallet/wallet2.cpp]: please use a define in the config for the hight

File: src/wallet/wallet2.cpp [hunk 2]
```diff
@@ -4097,10 +4104,16 @@ void wallet2::transfer_selected_rct(std::vector<cryptonote::tx_destination_entry
   }
   splitted_dsts.push_back(change_dts);
 
+  bool use_replay_prot = m_testnet || get_daemon_blockchain_height() >= 137510;
```
Comment (psychocrypt) [inline src/wallet/wallet2.cpp]: same: please use a define

File: src/wallet/wallet2.h [hunk 3]
```diff
@@ -1231,9 +1231,16 @@ namespace tools
       dust += d.amount;
     }
 
+    bool use_replay_prot = m_testnet || get_daemon_blockchain_height() >= 137510;
```
Comment (psychocrypt) [inline src/wallet/wallet2.h]: same: please use a define

File: src/cryptonote_core/blockchain.cpp [hunk 4]
```diff
@@ -2374,14 +2374,14 @@ bool Blockchain::check_tx_inputs(transaction& tx, tx_verification_context &tvc,
   
 
   // min/max tx version based on HF, and we accept v1 txes if having a non mixable
-  const size_t max_tx_version = CURRENT_TRANSACTION_VERSION;
+  const size_t max_tx_version = 3;
```
[the remark this answers is not in the document]
Comment (fireice-uk) [reply src/cryptonote_core/blockchain.cpp]: We are moving to a different base anyway, and for us both v2 and v3 will be valid so it was fairly meaningless.

File: src/cryptonote_core/cryptonote_basic.h [hunk 5]
```diff
@@ -1,375 +1,375 @@
...
+    uint64_t unlock_time;  //number of block (or time), used as a limitation like: spend this tx not early then block/time
+
+    std::vector<txin_v> vin;
+    std::vector<tx_out> vout;
+    //extra
+    std::vector<uint8_t> extra;
+
+    BEGIN_SERIALIZE()
+      VARINT_FIELD(version)
+      if(version == 0 || version > 3) return false;
```
[the remark this answers is not in the document]
Comment (fireice-uk) [reply src/cryptonote_core/cryptonote_basic.h]: Yes, it had windows line endings.

File: src/cryptonote_core/tx_pool.cpp [hunk 6]
```diff
@@ -599,10 +599,16 @@ namespace cryptonote
...
-  bool tx_memory_pool::fill_block_template(block &bl, size_t median_size, uint64_t already_generated_coins, size_t &total_size, uint64_t &fee, uint64_t height)
+  bool tx_memory_pool::fill_block_template(block &bl, size_t median_size, uint64_t already_generated_coins, size_t &total_size, uint64_t &fee, uint64_t height, bool testnet)
   {
     CRITICAL_REGION_LOCAL(m_transactions_lock);
 
+    bool allow_v3_tx;
+    if(testnet)
+      allow_v3_tx = height > 122670;
+    else
+      allow_v3_tx = height > 137501;
```
[the remark this answers is not in the document]
Comment (fireice-uk) [reply src/cryptonote_core/tx_pool.cpp]: There is no point. This is only temporary code to handle fork safely. We will do it properly in the new code.

File: src/wallet/wallet2.cpp (hunk 1 above)
Comment (fireice-uk) [reply src/wallet/wallet2.cpp]: There is no point. This is only temporary code to handle fork safely. We will do it properly in the new code.

File: src/cryptonote_core/cryptonote_format_utils.h [hunk 7]
```diff
@@ -95,8 +95,8 @@ namespace cryptonote
   };
 
   //---------------------------------------------------------------
-  bool construct_tx(const account_keys& sender_account_keys, const std::vector<tx_source_entry>& sources, const std::vector<tx_destination_entry>& destinations, std::vector<uint8_t> extra, transaction& tx, uint64_t unlock_time);
-  bool construct_tx_and_get_tx_key(const account_keys& sender_account_keys, const std::unordered_map<crypto::public_key, subaddress_index>& subaddresses, const std::vector<tx_source_entry>& sources, const std::vector<tx_destination_entry>& destinations, const cryptonote::account_public_address& change_addr, std::vector<uint8_t> extra, transaction& tx, uint64_t unlock_time, crypto::secret_key &tx_key, std::vector<crypto::secret_key> &additional_tx_keys, bool rct = true);
+ // bool construct_tx(const account_keys& sender_account_keys, const std::vector<tx_source_entry>& sources, const std::vector<tx_destination_entry>& destinations, std::vector<uint8_t> extra, transaction& tx, uint64_t unlock_time);
```
[the remark this answers is not in the document]
Comment (fireice-uk) [reply src/cryptonote_core/cryptonote_format_utils.h]: Ja wohl XD!

**Comment by psychocrypt** [comment]
Note: we will merge it without beautifying, this part will be cleaned in the next ryo code.

**Pull request merged**